servers/src/main/java/tachyon/master/lineage/meta/Lineage.java

```diff
@@ -0,0 +1,185 @@
+  }
+
+  public synchronized LineageInfo generateLineageInfo() {
+    LineageInfo info = new LineageInfo();
+    info.mId = mId;
+    List<Long> inputFiles = Lists.newArrayList();
+    for (TachyonFile file : mInputFiles) {
+      inputFiles.add(file.getFileId());
+    }
+    info.mInputFiles = inputFiles;
+
+    List<LineageFileInfo> outputFiles = Lists.newArrayList();
+    for (LineageFile lineageFile : mOutputFiles) {
+      outputFiles.add(lineageFile.generateLineageFileInfo());
+    }
+    info.mOutputFiles = outputFiles;
+
+    // TODO(yupeng) allow other types of jobs
+    info.mJob = ((CommandLineJob) mJob).generateCommandLineJobInfo();
+    info.mCreationTimeMs = mCreationTimeMs;
+    return info;
+  }
+
+  public synchronized List<TachyonFile> getInputFiles() {
+    return Collections.unmodifiableList(mInputFiles);
+  }
+
+  public synchronized List<LineageFile> getOutputFiles() {
+    return Collections.unmodifiableList(mOutputFiles);
+  }
+
+  public Job getJob() {
+    return mJob;
+  }
+
+  public long getId() {
+    return mId;
+  }
+
+  public long getCreationTime() {
+    return mCreationTimeMs;
+  }
+
+  public synchronized void updateOutputFileState(long fileId, LineageFileState newState) {
+    for (LineageFile outputFile : mOutputFiles) {
+      if (outputFile.getFileId() == fileId) {
+        outputFile.setState(newState);
+      }
+    }
+  }
+
+  public synchronized boolean needRecompute() {
+    for (LineageFile outputFile : mOutputFiles) {
+      if (outputFile.getState() == LineageFileState.LOST) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public synchronized boolean isCompleted() {
+    for (LineageFile outputFile : mOutputFiles) {
+      if (outputFile.getState() != LineageFileState.COMPLETED) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public synchronized boolean isPersisted() {
+    for (LineageFile outputFile : mOutputFiles) {
+      if (outputFile.getState() != LineageFileState.PERSISTED) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  public synchronized boolean isInCheckpointing() {
+    for (LineageFile outputFile : mOutputFiles) {
+      if (outputFile.getState() == LineageFileState.PERSISENCE_REQUESTED) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  public synchronized List<Long> getLostFiles() {
+    List<Long> result = Lists.newArrayList();
+    for (LineageFile outputFile : mOutputFiles) {
+      if (outputFile.getState() == LineageFileState.LOST) {
+        result.add(outputFile.getFileId());
+      }
+    }
+    return result;
+  }
+
+  public synchronized LineageFileState getOutputFileState(long fileId) {
+    for (LineageFile outputFile : mOutputFiles) {
+      if (outputFile.getFileId() == fileId) {
+        return outputFile.getState();
+      }
+    }
+    throw new RuntimeException("Output file " + fileId + " not found");
+  }
+
+  public synchronized LineageEntry toJournalEntry() {
```

Comment:
If you want to use this, I think you should implement "JournalEntryRepresentable"